### ISSUE TITLE
JUnit 5 Support

### ISSUE THREAD

**Issue by AustinShalit** [stefanbirkner/system-rules#55]
Just seeing if there are any plans to support JUnit 5 right now.

**Comment by stefanbirkner** [comment]
Yes, I create a JUnit Jupiter extension at the end of October.  I post updates here.

**Comment by kn9wBP2JX** [comment]
@stefanbirkner What can you suggest to use with JUnit 5 for stdout capturing?

**Comment by stefanbirkner** [comment]
Pleased wait for my JUnit Jupiter extension. I start development tomorrow.

**Comment by sbrannen** [comment]
> I start development tomorrow.

Awesome. Looking forward to it! 👍 

... b/c I otherwise might have to put something together myself. 😉

**Comment by sbrannen** [comment]
BTW, I've been meaning to show the Spring Boot team how to migrate their `OutputCapture` rule for JUnit 4 to JUnit Jupiter for a long time, so I just did it...

https://twitter.com/sam_brannen/status/919230917129981953

Not that you, @stefanbirkner, plan to do it the same way, but perhaps it serves as inspiration for you. 😉

**Comment by deepakab03** [comment]
Hello Stefan, 
I find the system-rules lib very useful and easy to use, thanks to you all! 
Any idea when you intend to release support for JUnit 5, so that we can do away with JUnit 4 as a dependency?

**Comment by stefanbirkner** [comment]
Currently I'm doing two things:

- Rewrite the tests to acceptance tests so that I can use this safety net for the migration.
- Migrate [Fake SFTP Server Rule](https://github.com/stefanbirkner/fake-sftp-server-rule) to [Fake SFTP Server Lambda](https://github.com/stefanbirkner/fake-sftp-server-lambda) which is independent from the test framework. Here I gain experience for the migration of System Rules.

**Comment by stefanbirkner** [comment]
The rewrite of the tests is finished. After fixing #61 I create a library that is independent of JUnit 4.

**Comment by syhily** [comment]
https://gist.github.com/syhily/1ae31ba0394668eb040917387a698449
Implemented a JUnit5 based ExpectedSystemExit

**Comment by davidkarlsen** [comment]
The lambda stuff for sftp is really nice, but sometimes the port is needed before at the class initialization level, so that it can be injected into systemproperties for spring-boot to pick it up before initialization.

**Comment by stefanbirkner** [comment]
I think you are talking about Fake SFTP Server Lambda. Can you create an issue [there](https://github.com/stefanbirkner/fake-sftp-server-lambda/issues) and describe what you are looking for.

**Comment by davidkarlsen** [comment]
Well - not really the lambda case. I've migrated from junit4 to junit5, see https://github.com/stefanbirkner/fake-sftp-server-rule/issues/11 for the initial issue.
Since junit4 rules are not junit5 compatible (except in the cases where the migration trick from junit is used - and that requires the junit-rules to be based on certain base-classes) I looked at using the lamba-based library instead - but that comes with it sets of problems, since I need the port - so that I can set this in a system property - so that the port is read by the spring-configuration - and that needs to happen before spring-test initializes.

**Comment by RonCrocker** [comment]
+1 for finishing up migrations to JUnit5, particularly for `EnvironmentVariables`

**Comment by stefanbirkner** [comment]
In the meantime you can copy code from https://github.com/stefanbirkner/system-lambda/blob/master/src/main/java/com/github/stefanbirkner/systemlambda/SystemLambda.java The replacement for EnvironmentVariables is the method `withEnvironmentVariable` together with the inner class `EnvironmentVariables`.

**Comment by cmathiesen** [comment]
Hello! Just wanted to chime in and mention that I've been working on updating our tests to JUnit5 and I tried using your EnvironmentVariables class from SystemLambda and it works very nicely! We're looking forward to a future release :)

**Comment by maguro** [comment]
> The replacement for EnvironmentVariables is the method `withEnvironmentVariable` together with the inner class `EnvironmentVariables`.

I'm sure I'm being dense.  How can I set the environmental variables once, for all tests to share?

**Comment by deeTEEcee** [comment]
i don't think you're being dense. it's pretty unclear to me because there's a link to something that has a method and inner class `WithEnvironmentVariable` and there's a file called `EnvironmentVariables.java` elsewhere. Basically, I don't know what to do here either :)

**Comment by michaelruocco** [comment]
@stefanbirkner thanks for the very useful library! I am keen for JUnit 5 support too, I like the look of the system-lambda library and would be happy to switch over when it is released, do you require any help in trying to get it ready for release? I would probably take a while to learn enough to provide anything useful but would happily try to help.

@deeTEEcee + @maguro just incase it helps I managed to make use of the system lambda code to set environment variables for my Junit 5 tests, these classes should help, but feel free to ask if anything is not clear?

I copied (and slightly adjusted) these two classes: https://github.com/michaelruocco/verification-context/tree/master/app/spring-app/src/test/java/uk/co/idv/app/environment. Then my tests that set the environment variables are here: https://github.com/michaelruocco/verification-context/blob/master/app/spring-app/src/test/java/uk/co/idv/app/config/MongoConnectionStringTest.java. I won't profess to understand all the code inside the WithEnvironmentVariables class does, but I understood enough to get it working!

**Comment by Purus** [comment]
> Hello! Just wanted to chime in and mention that I've been working on updating our tests to JUnit5 and I tried using your EnvironmentVariables class from SystemLambda and it works very nicely! We're looking forward to a future release :)

 Could you please share how you used it..

**Comment by amit-dd** [comment]
@michaelruocco Can you please check the links, they are 404'ing? This is for set environment variables for  Junit 5

**Comment by michaelruocco** [comment]
Hi @amit-dd, yeah you are quite right, the reason the links are 404ing is because I decided going with system properties was a simpler solution in the end. In case you are interested, the old links with the code I used for environment properties are below:

This was the class copied and tweaked from this library: https://github.com/michaelruocco/idv/blob/988cdfe45358764e8dc7e11d5efaa2daed5c769c/app/spring-app/src/test/java/uk/co/idv/app/environment/WithEnvironmentVariables.java

An example of this being used is here:
https://github.com/michaelruocco/idv/blob/ebbc91a05c4ef824b7cb458d9aa53cb1a57bb4ec/app/spring-app/src/test/java/uk/co/idv/app/ApplicationTest.java#L37

Hope that helps!

**Comment by vyvyd** [comment]
Hello Stefan, is JUnit5 support available?

**Comment by ashleyfrieze** [comment]
Jupiter support available in https://github.com/webcompere/system-stubs/tree/master/system-stubs-jupiter - a fork of `system-lambda`
